### PR TITLE
Add GPU filter operator (2D, 3D)

### PR DESCRIPTION
**Pull request by stiepan**
Signed-off-by: Kamil Tokarski <ktokarski@nvidia.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Adds GPU filter operator (2D and 3D convolutions with use-provided filters).

The relevant kernel part was merged here: https://github.com/NVIDIA/DALI/pull/4298

## Additional information:

1. A basic usage is just providing a batch of samples and batch of filters as two positional arguments (filters can reside on CPU or GPU).
2. By default the output has the same shape as the input, this means there is a need to handle somehow positions near the boundry that would cause OOB accesses. How it is done exactly is controlled with two parameters
  * `border` - what values to use to "pad" the input, possible modes are reflect 101, reflect 1001, wrap, clamp and constant. If the constant mode is used, user can provide a third argument that contains scalars to be used as a padding.
  * `anchor` - enables user to shift filter placement over the image
3. Alternatively, the output can be smaller so that there are no OOB accesses - this can be controlled with `mode` parameter which can be "same" (default) or "valid".

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2951

### CODE REVIEW

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
!build

File: dali/operators/image/convolution/filter.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+For video/sequence input, an array of scalars can be specified to be applied
+:func:`per-frame<nvidia.dali.fn.per_frame>`.
+
+.. note::
+  In fact, the operator computes a correlation, not a convolution,
+  i.e. the order of filter elements is not flipped when computing product of
+  the filter and the image.
+
+)code")
+    .NumInput(2, 3)
```
Comment (jantonguirao) [inline dali/operators/image/convolution/filter.cc:55]:
```suggestion
    .NumInput(2, 3)
    .InputDox(0, "data", "TensorList", R"code(Batch of input data.)code")
    .InputDox(1, "filter", "TensorList", R"code(Batch of filters.)code")
    .InputDox(2, ...)    
```
Please add documentation of the inputs, and perhaps remove the sentence in line 25.

File: dali/operators/image/convolution/filter.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+If the optional third positional argument is specified, it must be a batch of scalars.
+If ``"border"`` is set to ``"constant"``, the input samples will be padded with
+the corresponding scalars when convolved with the filter.
+The scalars must be of the same type as the input samples.
+For video/sequence input, an array of scalars can be specified to be applied
+:func:`per-frame<nvidia.dali.fn.per_frame>`.
+
+.. note::
+  In fact, the operator computes a correlation, not a convolution,
+  i.e. the order of filter elements is not flipped when computing product of
```
Comment (jantonguirao) [inline dali/operators/image/convolution/filter.cc:51]:
```suggestion
  i.e. the order of filter elements is not flipped when computing the product of
```

File: dali/operators/image/convolution/filter.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  the filter and the image.
+
+)code")
+    .NumInput(2, 3)
+    .NumOutput(1)
+    .AllowSequences()
+    .AddOptionalArg("anchor",
+                    R"code(Specifies position of the filter over the input.
```
Comment (jantonguirao) [inline dali/operators/image/convolution/filter.cc:59]:
```suggestion
                    R"code(Specifies the position of the filter over the input.
```

File: dali/operators/image/convolution/filter.cu [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#include <memory>
+
+#include "dali/operators/image/convolution/filter.h"
+#include "dali/operators/image/convolution/filter_gpu.h"
+
+namespace dali {
+
+namespace filter {
+
+extern template std::unique_ptr<OpImplBase<GPUBackend>>
```
Comment (jantonguirao) [inline dali/operators/image/convolution/filter.cu:24]: nitpick: use a macro to reduce verbosity?

File: dali/test/python/test_dali_variable_batch_size.py [hunk 5]
```diff
@@ -1232,6 +1232,43 @@ def sample_gen():
...
+def test_filter():
+
+    def filter_pipeline(max_batch_size, inputs, device):
+        batches = [list(zip(*batch)) for batch in inputs]
+        sample_batches = [list(inp_batch) for inp_batch, _, _ in batches]
+        filter_batches = [list(filt_batch) for _, filt_batch, _ in batches]
+        fill_value_bacthes = [list(fvs) for _, _, fvs in batches]
+
+        @pipeline_def
+        def piepline():
```
Comment (jantonguirao) [inline dali/test/python/test_dali_variable_batch_size.py:1244]:
```suggestion
        def pipeline():
```

File: dali/operators/image/convolution/filter.h [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  }
+
+  bool ShouldExpand(const Workspace& ws) override;
+
+  template <typename Out, typename In, typename W>
+  std::unique_ptr<OpImplBase<Backend>> GetFilterImpl(const OpSpec& spec_,
+                                                     const filter::InputDesc& input_desc);
+
+  bool SetupImpl(std::vector<OutputDesc>& output_desc, const Workspace& ws) override {
+    if (!impl_) {
```
Comment (jantonguirao) [inline dali/operators/image/convolution/filter.h:194]: how about checking that the current impl_ instance matches the previous input and filter type?

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/image/convolution/filter.h (hunk 6 above)
Comment (stiepan) [reply dali/operators/image/convolution/filter.h:194]: It makes sense on the one hand, on the other isn't it something that should the exectur do?

File: dali/test/python/test_dali_variable_batch_size.py (hunk 5 above)
Comment (stiepan) [reply dali/test/python/test_dali_variable_batch_size.py]: done

File: dali/operators/image/convolution/filter.cu (hunk 4 above)
Comment (stiepan) [reply dali/operators/image/convolution/filter.cu]: done

File: dali/operators/image/convolution/filter.cc (hunk 3 above)
Comment (stiepan) [reply dali/operators/image/convolution/filter.cc]: done

File: dali/operators/image/convolution/filter.cc (hunk 2 above)
Comment (stiepan) [reply dali/operators/image/convolution/filter.cc]: dine

File: dali/operators/image/convolution/filter.cc (hunk 1 above)
Comment (stiepan) [reply dali/operators/image/convolution/filter.cc]: Done, I moved whole sections from the main description to relevant parameters' docs.

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
I had to rebase over the operator tests split.

**Comment by stiepan** [comment]
!build

**Review by awolant — APPROVED** [review]

File: dali/operators/image/convolution/filter.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+within filter boundaries (``[0, ..., filter_extent - 1]``). The ordering of anchor's extents
+corresponds to the order of filter's extents.
+
+The parameter is ignored in ``"valid"`` mode.
+.)code",
+                    std::vector<int>{-1}, true, true)
+    .AddOptionalArg("border",
+                    R"code(Controls how to handle out-of-bound filter positions over the sample.
+
+Supported values are: ``"reflect_101"``, ``"reflect_1001"``, ``"wrap"``,
```
Comment (banasraf) [inline dali/operators/image/convolution/filter.cc:80]: In the tests I've seen just "101" and "1001" values and not "reflect_*". Is it something different?

Comment (stiepan) [reply dali/operators/image/convolution/filter.cc:80]: It's the same, just out of convenience. 

The boundary.h

```
if (type == "reflect_1001" || type == "reflect1001" || type == "1001")
```

**Review by banasraf — APPROVED** [review]

File: dali/operators/image/convolution/filter.cc (hunk 7 above)
Comment (banasraf) [reply dali/operators/image/convolution/filter.cc:80]: Ok, understood

**Pull request merged**